## Built-in Functions

### abs
Return the absolute value of an int or a float

In [1]:
abs(-12)

12

### divmod
return a pair of integers consisting of their integer quotient and remainder

In [2]:
divmod (7, 3)

(2, 1)

### min / max
Min / Max of non-empty sequence (list, string, tuple)

In [3]:
min([1, 3, 5, 7])

1

In [4]:
max('KPMG')

'P'

### round
Round to n digits after the decimal point

In [5]:
round(1.6666666666, 2)

1.67

Be aware of float precision!

In [6]:
c = 2.675
print(c)
print(round(c, 2))

2.675
2.67


In [7]:
2.675 == 2.67499999999999982236431605997495353221893310546875

True

***

### len
Size of an interable.

In [8]:
len("Hello World")

11

In [9]:
len([1, 2, 3])

3

### dir
Return a list of valid attributes for that object.

In [10]:
a_string = "Hello World"
dir(a_string)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',


In [11]:
a_string.__len__()

11

In [12]:
a_string.__hash__()

1816895075111600084

***

### filter(function, iterable)
Return a list of all elements of iterable where the fuction returns **True**

In [13]:
def my_filter_function(param):
    if isinstance(param, (int, float)):
        return True

In [14]:
worklist = [1, "2", "drei", 4, 5.5, 6.66]

In [15]:
filter(my_filter_function, worklist)

In [16]:
list(filter(my_filter_function, worklist))

[1, 4, 5.5, 6.66]

In [17]:
filter(lambda x: isinstance(x, (int, float)) , worklist)

In [18]:
list(filter(lambda x: isinstance(x, (int, float)) , worklist))

[1, 4, 5.5, 6.66]

### map(function, iterable)
Apply function to each item of a list

In [19]:
def my_map_function(param):
    if isinstance(param, (int, float)):
        return param ** 2
    return param

In [20]:
map(my_map_function, worklist)

In [21]:
list(map(my_map_function, worklist))

[1, '2', 'drei', 16, 30.25, 44.3556]

In [22]:
map(lambda x: x ** 2 if isinstance(x, (int, float)) else x, worklist)

In [23]:
list(map(lambda x: x ** 2 if isinstance(x, (int, float)) else x, worklist))

[1, '2', 'drei', 16, 30.25, 44.3556]

### reduce(function, iterable)
Apply function of two arguments cumulatively to the items of iterable, from left to right.


->> TODO Only applies to conseqcutive.

In [24]:
from functools import reduce

In [25]:
worklist = [1, 16, 30.25, 44.3556]

In [26]:
reduce(lambda x, y: x + y, worklist)

91.60560000000001

In [27]:
reduce(lambda x, y: x + y, worklist, 100.)

191.6056

***

### id(object)
The address of the object in memory

In [28]:
a = True
b = True
c = False
d = 10
e = .125

In [29]:
for _ in a, b, c, d, e:
    print(_, '\t', id(_))

True 	 4336104416
True 	 4336104416
False 	 4336105144
10 	 4483629584
0.125 	 4548046000


Note: a,b point to the same address (id) in memory, they both *refer* to 'True'

In [30]:
b = c

for _ in a, b, c, d, e:
    print(_, '\t', id(_))

True 	 4336104416
False 	 4336105144
False 	 4336105144
10 	 4483629584
0.125 	 4548046000


Note: c = point to the same address (id) in memory, they both *refer* to False.

In [31]:
c = .125
a = 10

for _ in a, b, c, d, e:
    print(_, '\t', id(_))

10 	 4483629584
False 	 4336105144
0.125 	 4548036464
10 	 4483629584
0.125 	 4548046000


In [32]:
id(a) == id(d)

True

In [33]:
id(c) == id(e)

False

Note: a and d both refer to the same **10** in memory.<br>
Though c and d not not refer to the same **0.125** in memory.<br>
Though the value of the tow is exactly the same.

In [34]:
c == e

True

In [ ]:
f = False
id(b) == id(f)

In [ ]:
g = 10
id(a) == id(d) == id(g)

Explaination:

* Python uses references, so there will always be only one **True** and **False** to be referred to in memory.
* **10** is handled exactly the same way, but **0.125** isn't. The reason is in the design: To inprove performance Python adds low integers to memory on startup, they will always remain regardless if referenced to.

***

### isinstance(object, classinfo)
Return if object is an instance of the classinfo argument

In [35]:
type("hello")

str

In [36]:
check_my_type = "Hello World!"

type(check_my_type)

str

In [37]:
isinstance(check_my_type, str)

True

In [38]:
isinstance(check_my_type, int)

False

In [39]:
isinstance(check_my_type, (int, str))

True

In [41]:
import datetime

check_my_type = datetime.datetime.utcnow()

In [42]:
isinstance(check_my_type, (int, str))

False

In [43]:
isinstance(check_my_type, datetime.datetime)

True

### callable

In [44]:
def my_function(): 
    return true

In [45]:
callable(my_function)

True

In [46]:
a_string = "Hello World"
callable(a_string)

False

***

### Enumerate

In [47]:
a_list = list("ABCDEFGHIJK")
a_list

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K']

#### First: some really bad ways to do it:

In [48]:
# worst
i = -1
for item in a_list:
    i += 1
    print(i, a_list[i])

0 A
1 B
2 C
3 D
4 E
5 F
6 G
7 H
8 I
9 J
10 K


In [49]:
# bad
i = 0
for item in a_list:
    print(i, a_list[i])
    i += 1

0 A
1 B
2 C
3 D
4 E
5 F
6 G
7 H
8 I
9 J
10 K


In [50]:
# acceptable
i = 0
for item in a_list:
    i += 1
    print(i, item)

1 A
2 B
3 C
4 D
5 E
6 F
7 G
8 H
9 I
10 J
11 K


#### The best way:

In [52]:
for i, item in enumerate(a_list):
    print(i, item)

0 A
1 B
2 C
3 D
4 E
5 F
6 G
7 H
8 I
9 J
10 K


In [56]:
for ei in enumerate(a_list):
    print(ei)

(0, 'A')
(1, 'B')
(2, 'C')
(3, 'D')
(4, 'E')
(5, 'F')
(6, 'G')
(7, 'H')
(8, 'I')
(9, 'J')
(10, 'K')


In [ ]:
a, b = (0, 'A')
print(a)
print(b)

* enumerate also accepts a start value

In [ ]:
for i, item in enumerate(a_list, start=11):
    print(i, item)

* enumerate also accepts generators

In [ ]:
for i, item in enumerate(range(10), start=11):
    print(i, item)

In [ ]:
def a_generator(n):
    i = 0
    while i < n:
        yield i
        i += 1
        
for i, item in enumerate(a_generator(10)):
    print(i, item)